storage/src/main/java/org/apache/uniffle/storage/factory/ShuffleHandlerFactory.java

```diff
@@ -126,6 +126,8 @@ private ClientReadHandler getMemoryClientReadHandler(
       Roaring64NavigableMap realExceptBlockIds = RssUtils.cloneBitMap(request.getExpectBlockIds());
       realExceptBlockIds.xor(request.getProcessBlockIds());
       expectTaskIds = RssUtils.generateTaskIdBitMap(realExceptBlockIds, request.getIdHelper());
+    } else {
+      expectTaskIds = request.getExpectTaskIds();
```

Comment:
Does this harm legacy Spark / TEZ / MR read clients?

Reply:
With `serverInfoList.size() == 1`, the `else` branch is used for TEZ and MR client. I think expected task ids and those task ids in block ids differ structurally for those clients. See related code in #1518.

I think with setting `.expectedTaskIdsBitmapFilterEnable(true)` for those two clients would fix this.